Handlers/AdvancedSKUManagementHandler.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Nwazet.Commerce.Models;
+using Nwazet.Commerce.Services;
+using Orchard;
+using Orchard.ContentManagement;
+using Orchard.ContentManagement.Handlers;
+using Orchard.Environment.Extensions;
+using Orchard.Localization;
+using Orchard.UI.Notify;
+
+namespace Nwazet.Commerce.Handlers {
+    [OrchardFeature("Nwazet.AdvancedSKUManagement")]
+    public class AdvancedSKUManagementHandler : ContentHandler {
+
+        private readonly IOrchardServices _orchardServices;
+        private readonly IContentManager _contentManager;
+        private readonly IEnumerable<ISKUUniquenessExceptionProvider> _SKUUniquenessExceptionProviders;
+        private readonly Lazy<ISKUGenerationServices> _SKUGenerationServices;
+
+        public AdvancedSKUManagementHandler(
+            IOrchardServices orchardServices,
+            IContentManager contentManager,
+            IEnumerable<ISKUUniquenessExceptionProvider> SKUUniquenessExceptionProviders,
+            Lazy<ISKUGenerationServices> SKUGenerationServices) {
+
+            _orchardServices = orchardServices;
+            _contentManager = contentManager;
+
+            _SKUUniquenessExceptionProviders = SKUUniquenessExceptionProviders;
+
+            _SKUGenerationServices = SKUGenerationServices;
+
+            T = NullLocalizer.Instance;
+
+            OnUpdated<ProductPart>((ctx, part) => CreateSku(part));
+            OnCreated<ProductPart>((ctx, part) => {
+                if (part.ContentItem.VersionRecord == null) {
+                    CreateSku(part);
+                }
+            });
+        }
+
+        public Localizer T { get; set; }
+
+        protected override void UpdateEditorShape(UpdateEditorContext context) {
+            base.UpdateEditorShape(context);
+            //Apply this only for ProductParts
+            if (context.ContentItem.Parts.Any(part => part is ProductPart)) {
+                var productPart = context.ContentItem.As<ProductPart>();
+                var settings = _orchardServices.WorkContext.CurrentSite.As<AdvancedSKUsSiteSettingPart>();
+                if (settings.RequireUniqueSKU) {
+                    var sameSKUProductIds = _contentManager.Query<ProductPart, ProductPartRecord>(VersionOptions.Latest)
+                        .Where(ppr => ppr.Id != productPart.Record.Id && ppr.Sku == productPart.Sku)
+                        .List().Select(pp => pp.Id);
```

Comment:
You're right. If that ever becomes a bottleneck however, we can change that into more efficient HQL.